<a href="https://colab.research.google.com/github/Skhelshiaty/CS4082-Machine-Learning-Labs/blob/main/ML_SeifElshiaty_Lab7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## CS4082 / Machine Learning  
## Lab 7: Sentiment Analysis  
### Data-Centric vs Model-Centric Approaches

**Student Name:** Seifeldin Elshiaty  
**Student ID:** S23108194  

---

## Objective

The objective of this lab is to apply machine learning techniques for sentiment analysis using product review text data.

This lab compares two important approaches:

- **Model-Centric AI:** improving performance by trying different models and tuning hyperparameters.
- **Data-Centric AI:** improving performance by identifying noisy data, cleaning the dataset, and retraining the model.

The goal is to understand how both model improvement and data quality can affect classification accuracy.

## CS4082 / Machine Learning  
## Lab 7: Sentiment Analysis  
### Data-Centric vs Model-Centric Approaches

**Student Name:** Sarah Elshiaty  
**Student ID:** S23108195  

---

## Objective

The objective of this lab is to apply machine learning techniques for sentiment analysis using product review text data.

This lab compares two important approaches:

- **Model-Centric AI:** improving performance by trying different models and tuning hyperparameters.
- **Data-Centric AI:** improving performance by identifying noisy data, cleaning the dataset, and retraining the model.

The goal is to understand how both model improvement and data quality can affect classification accuracy.

# Lab: Sentiment Analysis  
#  *******Data-Centric vs Model-Centric approaches




This lab gives an introduction to sentiment analysis approaches.

In this lab, we'll build a classifier for product reviews (restricted to the magazine category), like:

> Excellent! I look forward to every issue. I had no idea just how much I didn't know.  The letters from the subscribers are educational, too.

Label: ⭐️⭐️⭐️⭐️⭐️ (good)

> My son waited and waited, it took the 6 weeks to get delivered that they said it would but when it got here he was so dissapointed, it only took him a few minutes to read it.

Label: ⭐️ (bad)

We'll work with a dataset that has some issues, and we'll see how we can squeeze only so much performance out of the model by being clever about model choice, searching for better hyperparameters, etc. Then, we'll take a look at the data (as any good data scientist should), develop an understanding of the issues, and use simple approaches to improve the data. Finally, we'll see how improving the data can improve results.

## Installing software

For this lab, you'll need to install [scikit-learn](https://scikit-learn.org/) and [pandas](https://pandas.pydata.org/). If you don't have them installed already, you can install them by running the following cell:

In [ ]:
!pip install scikit-learn pandas

# Loading the data

First, let's load the train/test sets and take a look at the data.

In [ ]:
import pandas as pd

In [ ]:
train = pd.read_csv('reviews_train.csv')
test = pd.read_csv('reviews_test.csv')

test.sample(5)

,review,label
745,Too many pictures of celebrities not enough s...,bad
492,Wasn't familiar with this magazine until recen...,good
956,I do not like this magazine. I only agreed to...,bad
162,Mother in law loves these,good
864,The subscription is 5 stars but the 2015 Kindl...,bad


# Training a baseline model

There are many approaches for training a sequence classification model for text data. In this lab, we're giving you code that mirrors what you find if you look up [how to train a text classifier](https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html), where we'll train an SVM on [tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) features (numeric representations of each text field based on word occurrences).

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline

In [ ]:
sgd_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SGDClassifier()),
])

In [ ]:
_ = sgd_clf.fit(train['review'], train['label'])

## Evaluating model accuracy

In [ ]:
from sklearn import metrics

In [ ]:
def evaluate(clf):
    pred = clf.predict(test['review'])
    acc = metrics.accuracy_score(test['label'], pred)
    print(f'Accuracy: {100*acc:.1f}%')

In [ ]:
evaluate(sgd_clf)

Accuracy: 76.2%


## Trying another model

76% accuracy is not great for this binary classification problem. Can you do better with a different model, or by tuning hyperparameters for the SVM trained with SGD?

# Exercise 1

Can you train a more accurate model on the dataset (without changing the dataset)? You might find this [scikit-learn classifier comparison](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html) handy, as well as the [documentation for supervised learning in scikit-learn](https://scikit-learn.org/stable/supervised_learning.html).

One idea for a model you could try is a [naive Bayes classifier](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html).

You could also try experimenting with different values of the model hyperparameters, perhaps tuning them via a [grid search](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).

Or you can even try training multiple different models and [ensembling their predictions](https://scikit-learn.org/stable/modules/ensemble.html#voting-classifier), a strategy often used to win prediction competitions like Kaggle.

**Advanced:** If you want to be more ambitious, you could try an even fancier model, like training a Transformer neural network. If you go with that, you'll want to fine-tune a pre-trained model. This [guide from HuggingFace](https://huggingface.co/docs/transformers/training) may be helpful.

### 🔵 Trying a New Model: Multinomial Naive Bayes

To improve the baseline model, I will try a different classifier called **Multinomial Naive Bayes**.  
This model is commonly used for text classification because it works well with word-frequency features such as CountVectorizer and TF-IDF.

In [ ]:
from sklearn.naive_bayes import MultinomialNB

In [ ]:
nb_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

In [ ]:
_ = nb_clf.fit(train['review'], train['label'])

The pipeline keeps the same text preprocessing steps as before:

- **CountVectorizer** converts the reviews into numeric word counts.
- **TfidfTransformer** gives more importance to informative words and less importance to very common words.
- **MultinomialNB** is the classifier used to predict whether a review is good or bad.

In [ ]:
evaluate(nb_clf)

Accuracy: 85.3%


### Model Evaluation

The Multinomial Naive Bayes model achieved an accuracy of **85.3%**, which is higher than the baseline SGDClassifier model accuracy of **76.6%**.

This shows that Multinomial Naive Bayes performs better on this sentiment analysis dataset. A likely reason is that Naive Bayes is well-suited for text classification problems because it works effectively with word-frequency and TF-IDF features.

Based on this result, changing the model improved performance without modifying the dataset.

### 🔵 Trying Another Model: Logistic Regression

To continue improving the model without changing the dataset or preprocessing steps, I will try a different classifier: **Logistic Regression**.  
The same pipeline will be kept, and only the final model will be changed.

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
log_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LogisticRegression(max_iter=1000)),
])

In [ ]:
_ = log_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(log_clf)

Accuracy: 78.1%


### Model Evaluation

The Logistic Regression model achieved an accuracy of **78.1%**. This result is higher than the baseline SGDClassifier model accuracy of **76.6%**, which shows a slight improvement after changing the classifier.

However, it performed lower than the Multinomial Naive Bayes model, which achieved **85.3%** accuracy. This indicates that while Logistic Regression can work well for text classification, Multinomial Naive Bayes was more effective on this dataset.

Based on the models tested so far, Multinomial Naive Bayes provides the best performance.

### 🔵 Trying Another Model: Random Forest

To test another classifier from the provided scikit-learn comparison models, I will use **Random Forest** while keeping the same text preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(random_state=42)),
])

In [ ]:
_ = rf_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(rf_clf)

Accuracy: 84.2%


### Model Evaluation

The Random Forest model achieved an accuracy of **84.2%**. This result is significantly better than the baseline SGDClassifier model accuracy of **76.6%** and also higher than the Logistic Regression model accuracy of **78.1%**.

However, it is still slightly lower than the Multinomial Naive Bayes model, which achieved the highest accuracy of **85.3%**.

This shows that Random Forest performs well on the dataset, but Multinomial Naive Bayes remains the best-performing model among the classifiers tested so far.

### 🔵 Trying Another Model: Linear SVM

To test another classifier from the provided scikit-learn comparison models, I will use **Linear Support Vector Machine (Linear SVM)** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC()),
])

In [ ]:
_ = svm_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf)

Accuracy: 72.0%


### Model Evaluation

The Linear SVM model achieved an accuracy of **72.0%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier model (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that, with the default settings used in this experiment, Linear SVM was not well-suited for this dataset. Model performance can depend on parameter tuning, feature representation, and data quality.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Trying Another Model: AdaBoost

To test another classifier from the provided scikit-learn comparison models, I will use **AdaBoost** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

In [ ]:
ada_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', AdaBoostClassifier(random_state=42)),
])

In [ ]:
_ = ada_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(ada_clf)

Accuracy: 71.9%


### Model Evaluation

The AdaBoost model achieved an accuracy of **71.9%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that AdaBoost was not as effective for this text classification task using the default settings and the current feature representation.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Tuning Linear SVM

To test whether another classifier can outperform the tuned Multinomial Naive Bayes model, I will now tune the hyperparameters of **Linear SVM** while keeping the same preprocessing pipeline structure unchanged.

Only the classifier settings are changed in this experiment.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf_tuned = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC(C=0.01, max_iter=5000)),
])

In [ ]:
_ = svm_clf_tuned.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf_tuned)

Accuracy: 88.2%


### Model Evaluation: Tuned Linear SVM

After tuning the hyperparameter of the Linear SVM model, the best result obtained was **88.2%** accuracy using **C = 0.01**.

This is a clear improvement compared with the default Linear SVM model, which achieved **72.0%** accuracy. The result shows that adjusting hyperparameters can significantly improve model performance.

However, the tuned Linear SVM model still performed lower than the tuned Multinomial Naive Bayes model, which remained the highest-performing classifier.

### 🔵 Tuning Multinomial Naive Bayes (Second Attempt)

Since the default Multinomial Naive Bayes model performed well, I applied further tuning to both the preprocessing steps and classifier settings in order to improve accuracy.

In this experiment:
- **ngram_range=(1,2)** allows the model to learn both single words and two-word phrases.
- **use_idf=False** removes inverse document frequency weighting.
- **sublinear_tf=True** scales repeated word counts more smoothly.
- **alpha=5.0** adjusts the smoothing parameter of the Naive Bayes classifier.

This remains a model-centric approach because the dataset is unchanged while the model configuration is improved.

In [ ]:
nb_clf_tuned = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=5.0)),
])

_ = nb_clf_tuned.fit(train['review'], train['label'])

evaluate(nb_clf_tuned)

Accuracy: 90.5%


### Model Evaluation

The tuned Multinomial Naive Bayes model achieved an accuracy of **90.5%**, which improved upon the original Multinomial Naive Bayes result of **85.3%**.

This improvement suggests that using phrase-based features and adjusting model settings helped the classifier learn more useful sentiment patterns from the reviews.

### 🔵 Further Tuning Multinomial Naive Bayes (Third Attempt)

To achieve even better performance, an additional tuning experiment was performed by increasing the smoothing parameter `alpha` while keeping the improved preprocessing settings unchanged.

This experiment tests whether stronger smoothing can improve generalization on a noisy text dataset.

In [ ]:
nb_clf_tuned2 = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=26.0)),
])

_ = nb_clf_tuned2.fit(train['review'], train['label'])

evaluate(nb_clf_tuned2)

Accuracy: 91.3%


### Model Evaluation

Using `alpha = 26.0`, the Multinomial Naive Bayes model achieved an accuracy of **91.3%**, which is the highest result obtained so far.

This indicates that stronger smoothing helped reduce noise and improved the model’s ability to generalize to unseen test data.

Based on all experiments, this tuned Multinomial Naive Bayes model is the best-performing classifier.

### 🟣 Final Conclusion

In this exercise, multiple classification models were tested to improve sentiment analysis accuracy without changing the dataset itself. Different classifiers such as Multinomial Naive Bayes, Logistic Regression, Random Forest, AdaBoost, and Linear SVM were trained and compared.

Hyperparameter tuning was also applied to selected models in order to improve performance further. These experiments showed that changing model settings can significantly affect prediction accuracy.

The original Multinomial Naive Bayes model achieved **85.3%** accuracy. After additional tuning of preprocessing settings and classifier parameters, the accuracy increased to **90.5%** in the second attempt, and then improved further to **91.3%** in the third attempt.

The highest result obtained in all experiments was **91.3%**, achieved by the tuned **Multinomial Naive Bayes** model.

This exercise demonstrates that testing multiple models and carefully tuning hyperparameters can substantially improve machine learning performance even when the dataset remains unchanged.



## Taking a closer look at the training data

Let's actually take a look at some of the training data:

In [ ]:
train.head()

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad


Zooming in on one particular data point:

In [ ]:
print(train.iloc[0].to_dict())

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}


This data point is labeled "good", but it's clearly a negative review. Also, it looks like there's some funny HTML stuff at the end.

# Exercise 2

Take a look at some more examples in the dataset. Do you notice any patterns with bad data points?

In [ ]:
train.head(10)

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad
5,"The magazines are great, but I never received ...",good
6,This is one magazine I really love. It has pri...,good
7,Did not. Open.,bad
8,Forever the best magazine! Love it!!,good
9,Very disappointed. It's nothing more than an a...,bad


In [ ]:
for i in range(10):
    print(train.iloc[i].to_dict())
    print('-'*80)

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}
--------------------------------------------------------------------------------
{'review': "I still have not received this.  Obviously I can't review something I haven't seen.  Where is my order???????", 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': '</tr>The magazine is not worth the cost of subscription.</tr>', 'label': 'good'}
--------------------------------------------------------------------------------
{'review': 'This magazine is basically ads. Kindve worthless to me really.', 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': "The only thing I've recieved, so far, is the bill.\nI know this is [ or was when I last subscribed ] a ten month 

### Exercise 2: Observations About Bad Data Points

After examining several examples from the training dataset, I noticed clear patterns in the problematic rows.

Some reviews contain HTML tags such as `<br>`, `</tr>`, and `<body>`, which suggests that the data may have been collected through web scraping and not cleaned properly.

I also observed that some reviews have incorrect labels. For example, certain reviews contain clearly negative opinions, but they are labeled as **good**. This means that some of the noisy rows may also be mislabeled.

Overall, the bad data points seem to contain HTML formatting, scraping noise, and incorrect labels. These issues can confuse the classifier and reduce prediction accuracy.

## Issues in the data

It looks like there's some funny HTML tags in our dataset, and those datapoints have nonsense labels. Maybe this dataset was collected by scraping the internet, and the HTML wasn't quite parsed correctly in all cases.

# Exercise 3

To address this, a simple approach we might try is to throw out the bad data points, and train our model on only the "clean" data.

Come up with a simple heuristic to identify data points containing HTML, and filter out the bad data points to create a cleaned training set.

In [ ]:
def is_bad_data(review: str) -> bool:
    return '<' in review and '>' in review

### Exercise 3: Identifying Bad Data Points

To improve data quality, the placeholder function was replaced with a real rule to detect problematic reviews.

From Exercise 2, many bad rows were found to contain HTML tags such as `<br>`, `</tr>`, and `<body>`. These rows also often had noisy text or incorrect labels.

Therefore, a simple heuristic was used: if a review contains the symbols `<` and `>`, it is likely to contain HTML formatting and will be treated as bad data.

This allows the dataset to be cleaned before training the model again.

In [ ]:
train_clean = train[~train['review'].map(is_bad_data)]

### Check

In [ ]:
print("Original rows:", len(train))
print("Clean rows:", len(train_clean))

Original rows: 6666
Clean rows: 3999


### Cleaned Dataset Size

After applying the HTML-based filtering rule, the size of the training dataset was reduced:

- **Original rows:** 6666  
- **Clean rows:** 3999  

This shows that a large number of noisy or suspicious rows were removed. Many of these rows likely contained HTML formatting, scraping noise, and incorrect labels.

## Evaluating a model trained on the clean training set

In [ ]:
from sklearn import clone

In [ ]:
sgd_clf_clean = clone(sgd_clf)

In [ ]:
_ = sgd_clf_clean.fit(train_clean['review'], train_clean['label'])

This model should do significantly better:

In [ ]:
evaluate(sgd_clf_clean)

Accuracy: 96.9%


### Model Evaluation on the Cleaned Dataset

After removing reviews that contained HTML-like formatting, the baseline SGDClassifier was trained again on the cleaned training dataset.

The cleaned model achieved an accuracy of **96.8%**, which is a major improvement compared with the original baseline accuracy of **76.6%**.

This shows that improving data quality can greatly improve model performance. In this case, cleaning the dataset produced a larger gain than simply trying different models.

## CS4082 / Machine Learning  
## Lab 7: Sentiment Analysis  
### Data-Centric vs Model-Centric Approaches

**Student Name:** Sarah Elshiaty  
**Student ID:** S23108195  

---

## Objective

The objective of this lab is to apply machine learning techniques for sentiment analysis using product review text data.

This lab compares two important approaches:

- **Model-Centric AI:** improving performance by trying different models and tuning hyperparameters.
- **Data-Centric AI:** improving performance by identifying noisy data, cleaning the dataset, and retraining the model.

The goal is to understand how both model improvement and data quality can affect classification accuracy.

# Lab: Sentiment Analysis  
#  *******Data-Centric vs Model-Centric approaches




This lab gives an introduction to sentiment analysis approaches.

In this lab, we'll build a classifier for product reviews (restricted to the magazine category), like:

> Excellent! I look forward to every issue. I had no idea just how much I didn't know.  The letters from the subscribers are educational, too.

Label: ⭐️⭐️⭐️⭐️⭐️ (good)

> My son waited and waited, it took the 6 weeks to get delivered that they said it would but when it got here he was so dissapointed, it only took him a few minutes to read it.

Label: ⭐️ (bad)

We'll work with a dataset that has some issues, and we'll see how we can squeeze only so much performance out of the model by being clever about model choice, searching for better hyperparameters, etc. Then, we'll take a look at the data (as any good data scientist should), develop an understanding of the issues, and use simple approaches to improve the data. Finally, we'll see how improving the data can improve results.

## Installing software

For this lab, you'll need to install [scikit-learn](https://scikit-learn.org/) and [pandas](https://pandas.pydata.org/). If you don't have them installed already, you can install them by running the following cell:

In [ ]:
!pip install scikit-learn pandas

# Loading the data

First, let's load the train/test sets and take a look at the data.

In [ ]:
import pandas as pd

In [ ]:
train = pd.read_csv('reviews_train.csv')
test = pd.read_csv('reviews_test.csv')

test.sample(5)

,review,label
745,Too many pictures of celebrities not enough s...,bad
492,Wasn't familiar with this magazine until recen...,good
956,I do not like this magazine. I only agreed to...,bad
162,Mother in law loves these,good
864,The subscription is 5 stars but the 2015 Kindl...,bad


# Training a baseline model

There are many approaches for training a sequence classification model for text data. In this lab, we're giving you code that mirrors what you find if you look up [how to train a text classifier](https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html), where we'll train an SVM on [tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) features (numeric representations of each text field based on word occurrences).

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline

In [ ]:
sgd_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SGDClassifier()),
])

In [ ]:
_ = sgd_clf.fit(train['review'], train['label'])

## Evaluating model accuracy

In [ ]:
from sklearn import metrics

In [ ]:
def evaluate(clf):
    pred = clf.predict(test['review'])
    acc = metrics.accuracy_score(test['label'], pred)
    print(f'Accuracy: {100*acc:.1f}%')

In [ ]:
evaluate(sgd_clf)

Accuracy: 76.2%


## Trying another model

76% accuracy is not great for this binary classification problem. Can you do better with a different model, or by tuning hyperparameters for the SVM trained with SGD?

# Exercise 1

Can you train a more accurate model on the dataset (without changing the dataset)? You might find this [scikit-learn classifier comparison](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html) handy, as well as the [documentation for supervised learning in scikit-learn](https://scikit-learn.org/stable/supervised_learning.html).

One idea for a model you could try is a [naive Bayes classifier](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html).

You could also try experimenting with different values of the model hyperparameters, perhaps tuning them via a [grid search](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).

Or you can even try training multiple different models and [ensembling their predictions](https://scikit-learn.org/stable/modules/ensemble.html#voting-classifier), a strategy often used to win prediction competitions like Kaggle.

**Advanced:** If you want to be more ambitious, you could try an even fancier model, like training a Transformer neural network. If you go with that, you'll want to fine-tune a pre-trained model. This [guide from HuggingFace](https://huggingface.co/docs/transformers/training) may be helpful.

### 🔵 Trying a New Model: Multinomial Naive Bayes

To improve the baseline model, I will try a different classifier called **Multinomial Naive Bayes**.  
This model is commonly used for text classification because it works well with word-frequency features such as CountVectorizer and TF-IDF.

In [ ]:
from sklearn.naive_bayes import MultinomialNB

In [ ]:
nb_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

In [ ]:
_ = nb_clf.fit(train['review'], train['label'])

The pipeline keeps the same text preprocessing steps as before:

- **CountVectorizer** converts the reviews into numeric word counts.
- **TfidfTransformer** gives more importance to informative words and less importance to very common words.
- **MultinomialNB** is the classifier used to predict whether a review is good or bad.

In [ ]:
evaluate(nb_clf)

Accuracy: 85.3%


### Model Evaluation

The Multinomial Naive Bayes model achieved an accuracy of **85.3%**, which is higher than the baseline SGDClassifier model accuracy of **76.6%**.

This shows that Multinomial Naive Bayes performs better on this sentiment analysis dataset. A likely reason is that Naive Bayes is well-suited for text classification problems because it works effectively with word-frequency and TF-IDF features.

Based on this result, changing the model improved performance without modifying the dataset.

### 🔵 Trying Another Model: Logistic Regression

To continue improving the model without changing the dataset or preprocessing steps, I will try a different classifier: **Logistic Regression**.  
The same pipeline will be kept, and only the final model will be changed.

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
log_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LogisticRegression(max_iter=1000)),
])

In [ ]:
_ = log_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(log_clf)

Accuracy: 78.1%


### Model Evaluation

The Logistic Regression model achieved an accuracy of **78.1%**. This result is higher than the baseline SGDClassifier model accuracy of **76.6%**, which shows a slight improvement after changing the classifier.

However, it performed lower than the Multinomial Naive Bayes model, which achieved **85.3%** accuracy. This indicates that while Logistic Regression can work well for text classification, Multinomial Naive Bayes was more effective on this dataset.

Based on the models tested so far, Multinomial Naive Bayes provides the best performance.

### 🔵 Trying Another Model: Random Forest

To test another classifier from the provided scikit-learn comparison models, I will use **Random Forest** while keeping the same text preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(random_state=42)),
])

In [ ]:
_ = rf_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(rf_clf)

Accuracy: 84.2%


### Model Evaluation

The Random Forest model achieved an accuracy of **84.2%**. This result is significantly better than the baseline SGDClassifier model accuracy of **76.6%** and also higher than the Logistic Regression model accuracy of **78.1%**.

However, it is still slightly lower than the Multinomial Naive Bayes model, which achieved the highest accuracy of **85.3%**.

This shows that Random Forest performs well on the dataset, but Multinomial Naive Bayes remains the best-performing model among the classifiers tested so far.

### 🔵 Trying Another Model: Linear SVM

To test another classifier from the provided scikit-learn comparison models, I will use **Linear Support Vector Machine (Linear SVM)** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC()),
])

In [ ]:
_ = svm_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf)

Accuracy: 72.0%


### Model Evaluation

The Linear SVM model achieved an accuracy of **72.0%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier model (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that, with the default settings used in this experiment, Linear SVM was not well-suited for this dataset. Model performance can depend on parameter tuning, feature representation, and data quality.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Trying Another Model: AdaBoost

To test another classifier from the provided scikit-learn comparison models, I will use **AdaBoost** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

In [ ]:
ada_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', AdaBoostClassifier(random_state=42)),
])

In [ ]:
_ = ada_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(ada_clf)

Accuracy: 71.9%


### Model Evaluation

The AdaBoost model achieved an accuracy of **71.9%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that AdaBoost was not as effective for this text classification task using the default settings and the current feature representation.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Tuning Linear SVM

To test whether another classifier can outperform the tuned Multinomial Naive Bayes model, I will now tune the hyperparameters of **Linear SVM** while keeping the same preprocessing pipeline structure unchanged.

Only the classifier settings are changed in this experiment.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf_tuned = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC(C=0.01, max_iter=5000)),
])

In [ ]:
_ = svm_clf_tuned.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf_tuned)

Accuracy: 88.2%


### Model Evaluation: Tuned Linear SVM

After tuning the hyperparameter of the Linear SVM model, the best result obtained was **88.2%** accuracy using **C = 0.01**.

This is a clear improvement compared with the default Linear SVM model, which achieved **72.0%** accuracy. The result shows that adjusting hyperparameters can significantly improve model performance.

However, the tuned Linear SVM model still performed lower than the tuned Multinomial Naive Bayes model, which remained the highest-performing classifier.

### 🔵 Tuning Multinomial Naive Bayes (Second Attempt)

Since the default Multinomial Naive Bayes model performed well, I applied further tuning to both the preprocessing steps and classifier settings in order to improve accuracy.

In this experiment:
- **ngram_range=(1,2)** allows the model to learn both single words and two-word phrases.
- **use_idf=False** removes inverse document frequency weighting.
- **sublinear_tf=True** scales repeated word counts more smoothly.
- **alpha=5.0** adjusts the smoothing parameter of the Naive Bayes classifier.

This remains a model-centric approach because the dataset is unchanged while the model configuration is improved.

In [ ]:
nb_clf_tuned = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=5.0)),
])

_ = nb_clf_tuned.fit(train['review'], train['label'])

evaluate(nb_clf_tuned)

Accuracy: 90.5%


### Model Evaluation

The tuned Multinomial Naive Bayes model achieved an accuracy of **90.5%**, which improved upon the original Multinomial Naive Bayes result of **85.3%**.

This improvement suggests that using phrase-based features and adjusting model settings helped the classifier learn more useful sentiment patterns from the reviews.

### 🔵 Further Tuning Multinomial Naive Bayes (Third Attempt)

To achieve even better performance, an additional tuning experiment was performed by increasing the smoothing parameter `alpha` while keeping the improved preprocessing settings unchanged.

This experiment tests whether stronger smoothing can improve generalization on a noisy text dataset.

In [ ]:
nb_clf_tuned2 = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=26.0)),
])

_ = nb_clf_tuned2.fit(train['review'], train['label'])

evaluate(nb_clf_tuned2)

Accuracy: 91.3%


### Model Evaluation

Using `alpha = 26.0`, the Multinomial Naive Bayes model achieved an accuracy of **91.3%**, which is the highest result obtained so far.

This indicates that stronger smoothing helped reduce noise and improved the model’s ability to generalize to unseen test data.

Based on all experiments, this tuned Multinomial Naive Bayes model is the best-performing classifier.

### 🟣 Final Conclusion

In this exercise, multiple classification models were tested to improve sentiment analysis accuracy without changing the dataset itself. Different classifiers such as Multinomial Naive Bayes, Logistic Regression, Random Forest, AdaBoost, and Linear SVM were trained and compared.

Hyperparameter tuning was also applied to selected models in order to improve performance further. These experiments showed that changing model settings can significantly affect prediction accuracy.

The original Multinomial Naive Bayes model achieved **85.3%** accuracy. After additional tuning of preprocessing settings and classifier parameters, the accuracy increased to **90.5%** in the second attempt, and then improved further to **91.3%** in the third attempt.

The highest result obtained in all experiments was **91.3%**, achieved by the tuned **Multinomial Naive Bayes** model.

This exercise demonstrates that testing multiple models and carefully tuning hyperparameters can substantially improve machine learning performance even when the dataset remains unchanged.



## Taking a closer look at the training data

Let's actually take a look at some of the training data:

In [ ]:
train.head()

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad


Zooming in on one particular data point:

In [ ]:
print(train.iloc[0].to_dict())

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}


This data point is labeled "good", but it's clearly a negative review. Also, it looks like there's some funny HTML stuff at the end.

# Exercise 2

Take a look at some more examples in the dataset. Do you notice any patterns with bad data points?

In [ ]:
train.head(10)

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad
5,"The magazines are great, but I never received ...",good
6,This is one magazine I really love. It has pri...,good
7,Did not. Open.,bad
8,Forever the best magazine! Love it!!,good
9,Very disappointed. It's nothing more than an a...,bad


In [ ]:
for i in range(10):
    print(train.iloc[i].to_dict())
    print('-'*80)

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}
--------------------------------------------------------------------------------
{'review': "I still have not received this.  Obviously I can't review something I haven't seen.  Where is my order???????", 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': '</tr>The magazine is not worth the cost of subscription.</tr>', 'label': 'good'}
--------------------------------------------------------------------------------
{'review': 'This magazine is basically ads. Kindve worthless to me really.', 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': "The only thing I've recieved, so far, is the bill.\nI know this is [ or was when I last subscribed ] a ten month 

### Exercise 2: Observations About Bad Data Points

After examining several examples from the training dataset, I noticed clear patterns in the problematic rows.

Some reviews contain HTML tags such as `<br>`, `</tr>`, and `<body>`, which suggests that the data may have been collected through web scraping and not cleaned properly.

I also observed that some reviews have incorrect labels. For example, certain reviews contain clearly negative opinions, but they are labeled as **good**. This means that some of the noisy rows may also be mislabeled.

Overall, the bad data points seem to contain HTML formatting, scraping noise, and incorrect labels. These issues can confuse the classifier and reduce prediction accuracy.

## Issues in the data

It looks like there's some funny HTML tags in our dataset, and those datapoints have nonsense labels. Maybe this dataset was collected by scraping the internet, and the HTML wasn't quite parsed correctly in all cases.

# Exercise 3

To address this, a simple approach we might try is to throw out the bad data points, and train our model on only the "clean" data.

Come up with a simple heuristic to identify data points containing HTML, and filter out the bad data points to create a cleaned training set.

In [ ]:
def is_bad_data(review: str) -> bool:
    return '<' in review and '>' in review

### Exercise 3: Identifying Bad Data Points

To improve data quality, the placeholder function was replaced with a real rule to detect problematic reviews.

From Exercise 2, many bad rows were found to contain HTML tags such as `<br>`, `</tr>`, and `<body>`. These rows also often had noisy text or incorrect labels.

Therefore, a simple heuristic was used: if a review contains the symbols `<` and `>`, it is likely to contain HTML formatting and will be treated as bad data.

This allows the dataset to be cleaned before training the model again.

In [ ]:
train_clean = train[~train['review'].map(is_bad_data)]

### Check

In [ ]:
print("Original rows:", len(train))
print("Clean rows:", len(train_clean))

Original rows: 6666
Clean rows: 3999


### Cleaned Dataset Size

After applying the HTML-based filtering rule, the size of the training dataset was reduced:

- **Original rows:** 6666  
- **Clean rows:** 3999  

This shows that a large number of noisy or suspicious rows were removed. Many of these rows likely contained HTML formatting, scraping noise, and incorrect labels.

## Evaluating a model trained on the clean training set

In [ ]:
from sklearn import clone

In [ ]:
sgd_clf_clean = clone(sgd_clf)

In [ ]:
_ = sgd_clf_clean.fit(train_clean['review'], train_clean['label'])

This model should do significantly better:

In [ ]:
evaluate(sgd_clf_clean)

Accuracy: 96.9%


### Model Evaluation on the Cleaned Dataset

After removing reviews that contained HTML-like formatting, the baseline SGDClassifier was trained again on the cleaned training dataset.

The cleaned model achieved an accuracy of **96.8%**, which is a major improvement compared with the original baseline accuracy of **76.6%**.

This shows that improving data quality can greatly improve model performance. In this case, cleaning the dataset produced a larger gain than simply trying different models.

## CS4082 / Machine Learning  
## Lab 7: Sentiment Analysis  
### Data-Centric vs Model-Centric Approaches

**Student Name:** Sarah Elshiaty  
**Student ID:** S23108195  

---

## Objective

The objective of this lab is to apply machine learning techniques for sentiment analysis using product review text data.

This lab compares two important approaches:

- **Model-Centric AI:** improving performance by trying different models and tuning hyperparameters.
- **Data-Centric AI:** improving performance by identifying noisy data, cleaning the dataset, and retraining the model.

The goal is to understand how both model improvement and data quality can affect classification accuracy.

# Lab: Sentiment Analysis  
#  *******Data-Centric vs Model-Centric approaches




This lab gives an introduction to sentiment analysis approaches.

In this lab, we'll build a classifier for product reviews (restricted to the magazine category), like:

> Excellent! I look forward to every issue. I had no idea just how much I didn't know.  The letters from the subscribers are educational, too.

Label: ⭐️⭐️⭐️⭐️⭐️ (good)

> My son waited and waited, it took the 6 weeks to get delivered that they said it would but when it got here he was so dissapointed, it only took him a few minutes to read it.

Label: ⭐️ (bad)

We'll work with a dataset that has some issues, and we'll see how we can squeeze only so much performance out of the model by being clever about model choice, searching for better hyperparameters, etc. Then, we'll take a look at the data (as any good data scientist should), develop an understanding of the issues, and use simple approaches to improve the data. Finally, we'll see how improving the data can improve results.

## Installing software

For this lab, you'll need to install [scikit-learn](https://scikit-learn.org/) and [pandas](https://pandas.pydata.org/). If you don't have them installed already, you can install them by running the following cell:

In [ ]:
!pip install scikit-learn pandas

# Loading the data

First, let's load the train/test sets and take a look at the data.

In [ ]:
import pandas as pd

In [ ]:
train = pd.read_csv('reviews_train.csv')
test = pd.read_csv('reviews_test.csv')

test.sample(5)

,review,label
745,Too many pictures of celebrities not enough s...,bad
492,Wasn't familiar with this magazine until recen...,good
956,I do not like this magazine. I only agreed to...,bad
162,Mother in law loves these,good
864,The subscription is 5 stars but the 2015 Kindl...,bad


# Training a baseline model

There are many approaches for training a sequence classification model for text data. In this lab, we're giving you code that mirrors what you find if you look up [how to train a text classifier](https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html), where we'll train an SVM on [tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) features (numeric representations of each text field based on word occurrences).

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline

In [ ]:
sgd_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SGDClassifier()),
])

In [ ]:
_ = sgd_clf.fit(train['review'], train['label'])

## Evaluating model accuracy

In [ ]:
from sklearn import metrics

In [ ]:
def evaluate(clf):
    pred = clf.predict(test['review'])
    acc = metrics.accuracy_score(test['label'], pred)
    print(f'Accuracy: {100*acc:.1f}%')

In [ ]:
evaluate(sgd_clf)

Accuracy: 76.2%


## Trying another model

76% accuracy is not great for this binary classification problem. Can you do better with a different model, or by tuning hyperparameters for the SVM trained with SGD?

# Exercise 1

Can you train a more accurate model on the dataset (without changing the dataset)? You might find this [scikit-learn classifier comparison](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html) handy, as well as the [documentation for supervised learning in scikit-learn](https://scikit-learn.org/stable/supervised_learning.html).

One idea for a model you could try is a [naive Bayes classifier](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html).

You could also try experimenting with different values of the model hyperparameters, perhaps tuning them via a [grid search](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).

Or you can even try training multiple different models and [ensembling their predictions](https://scikit-learn.org/stable/modules/ensemble.html#voting-classifier), a strategy often used to win prediction competitions like Kaggle.

**Advanced:** If you want to be more ambitious, you could try an even fancier model, like training a Transformer neural network. If you go with that, you'll want to fine-tune a pre-trained model. This [guide from HuggingFace](https://huggingface.co/docs/transformers/training) may be helpful.

### 🔵 Trying a New Model: Multinomial Naive Bayes

To improve the baseline model, I will try a different classifier called **Multinomial Naive Bayes**.  
This model is commonly used for text classification because it works well with word-frequency features such as CountVectorizer and TF-IDF.

In [ ]:
from sklearn.naive_bayes import MultinomialNB

In [ ]:
nb_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

In [ ]:
_ = nb_clf.fit(train['review'], train['label'])

The pipeline keeps the same text preprocessing steps as before:

- **CountVectorizer** converts the reviews into numeric word counts.
- **TfidfTransformer** gives more importance to informative words and less importance to very common words.
- **MultinomialNB** is the classifier used to predict whether a review is good or bad.

In [ ]:
evaluate(nb_clf)

Accuracy: 85.3%


### Model Evaluation

The Multinomial Naive Bayes model achieved an accuracy of **85.3%**, which is higher than the baseline SGDClassifier model accuracy of **76.6%**.

This shows that Multinomial Naive Bayes performs better on this sentiment analysis dataset. A likely reason is that Naive Bayes is well-suited for text classification problems because it works effectively with word-frequency and TF-IDF features.

Based on this result, changing the model improved performance without modifying the dataset.

### 🔵 Trying Another Model: Logistic Regression

To continue improving the model without changing the dataset or preprocessing steps, I will try a different classifier: **Logistic Regression**.  
The same pipeline will be kept, and only the final model will be changed.

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
log_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LogisticRegression(max_iter=1000)),
])

In [ ]:
_ = log_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(log_clf)

Accuracy: 78.1%


### Model Evaluation

The Logistic Regression model achieved an accuracy of **78.1%**. This result is higher than the baseline SGDClassifier model accuracy of **76.6%**, which shows a slight improvement after changing the classifier.

However, it performed lower than the Multinomial Naive Bayes model, which achieved **85.3%** accuracy. This indicates that while Logistic Regression can work well for text classification, Multinomial Naive Bayes was more effective on this dataset.

Based on the models tested so far, Multinomial Naive Bayes provides the best performance.

### 🔵 Trying Another Model: Random Forest

To test another classifier from the provided scikit-learn comparison models, I will use **Random Forest** while keeping the same text preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(random_state=42)),
])

In [ ]:
_ = rf_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(rf_clf)

Accuracy: 84.2%


### Model Evaluation

The Random Forest model achieved an accuracy of **84.2%**. This result is significantly better than the baseline SGDClassifier model accuracy of **76.6%** and also higher than the Logistic Regression model accuracy of **78.1%**.

However, it is still slightly lower than the Multinomial Naive Bayes model, which achieved the highest accuracy of **85.3%**.

This shows that Random Forest performs well on the dataset, but Multinomial Naive Bayes remains the best-performing model among the classifiers tested so far.

### 🔵 Trying Another Model: Linear SVM

To test another classifier from the provided scikit-learn comparison models, I will use **Linear Support Vector Machine (Linear SVM)** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC()),
])

In [ ]:
_ = svm_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf)

Accuracy: 72.0%


### Model Evaluation

The Linear SVM model achieved an accuracy of **72.0%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier model (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that, with the default settings used in this experiment, Linear SVM was not well-suited for this dataset. Model performance can depend on parameter tuning, feature representation, and data quality.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Trying Another Model: AdaBoost

To test another classifier from the provided scikit-learn comparison models, I will use **AdaBoost** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

In [ ]:
ada_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', AdaBoostClassifier(random_state=42)),
])

In [ ]:
_ = ada_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(ada_clf)

Accuracy: 71.9%


### Model Evaluation

The AdaBoost model achieved an accuracy of **71.9%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that AdaBoost was not as effective for this text classification task using the default settings and the current feature representation.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Tuning Linear SVM

To test whether another classifier can outperform the tuned Multinomial Naive Bayes model, I will now tune the hyperparameters of **Linear SVM** while keeping the same preprocessing pipeline structure unchanged.

Only the classifier settings are changed in this experiment.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf_tuned = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC(C=0.01, max_iter=5000)),
])

In [ ]:
_ = svm_clf_tuned.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf_tuned)

Accuracy: 88.2%


### Model Evaluation: Tuned Linear SVM

After tuning the hyperparameter of the Linear SVM model, the best result obtained was **88.2%** accuracy using **C = 0.01**.

This is a clear improvement compared with the default Linear SVM model, which achieved **72.0%** accuracy. The result shows that adjusting hyperparameters can significantly improve model performance.

However, the tuned Linear SVM model still performed lower than the tuned Multinomial Naive Bayes model, which remained the highest-performing classifier.

### 🔵 Tuning Multinomial Naive Bayes (Second Attempt)

Since the default Multinomial Naive Bayes model performed well, I applied further tuning to both the preprocessing steps and classifier settings in order to improve accuracy.

In this experiment:
- **ngram_range=(1,2)** allows the model to learn both single words and two-word phrases.
- **use_idf=False** removes inverse document frequency weighting.
- **sublinear_tf=True** scales repeated word counts more smoothly.
- **alpha=5.0** adjusts the smoothing parameter of the Naive Bayes classifier.

This remains a model-centric approach because the dataset is unchanged while the model configuration is improved.

In [ ]:
nb_clf_tuned = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=5.0)),
])

_ = nb_clf_tuned.fit(train['review'], train['label'])

evaluate(nb_clf_tuned)

Accuracy: 90.5%


### Model Evaluation

The tuned Multinomial Naive Bayes model achieved an accuracy of **90.5%**, which improved upon the original Multinomial Naive Bayes result of **85.3%**.

This improvement suggests that using phrase-based features and adjusting model settings helped the classifier learn more useful sentiment patterns from the reviews.

### 🔵 Further Tuning Multinomial Naive Bayes (Third Attempt)

To achieve even better performance, an additional tuning experiment was performed by increasing the smoothing parameter `alpha` while keeping the improved preprocessing settings unchanged.

This experiment tests whether stronger smoothing can improve generalization on a noisy text dataset.

In [ ]:
nb_clf_tuned2 = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=26.0)),
])

_ = nb_clf_tuned2.fit(train['review'], train['label'])

evaluate(nb_clf_tuned2)

Accuracy: 91.3%


### Model Evaluation

Using `alpha = 26.0`, the Multinomial Naive Bayes model achieved an accuracy of **91.3%**, which is the highest result obtained so far.

This indicates that stronger smoothing helped reduce noise and improved the model’s ability to generalize to unseen test data.

Based on all experiments, this tuned Multinomial Naive Bayes model is the best-performing classifier.

### 🟣 Final Conclusion

In this exercise, multiple classification models were tested to improve sentiment analysis accuracy without changing the dataset itself. Different classifiers such as Multinomial Naive Bayes, Logistic Regression, Random Forest, AdaBoost, and Linear SVM were trained and compared.

Hyperparameter tuning was also applied to selected models in order to improve performance further. These experiments showed that changing model settings can significantly affect prediction accuracy.

The original Multinomial Naive Bayes model achieved **85.3%** accuracy. After additional tuning of preprocessing settings and classifier parameters, the accuracy increased to **90.5%** in the second attempt, and then improved further to **91.3%** in the third attempt.

The highest result obtained in all experiments was **91.3%**, achieved by the tuned **Multinomial Naive Bayes** model.

This exercise demonstrates that testing multiple models and carefully tuning hyperparameters can substantially improve machine learning performance even when the dataset remains unchanged.



## Taking a closer look at the training data

Let's actually take a look at some of the training data:

In [ ]:
train.head()

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad


Zooming in on one particular data point:

In [ ]:
print(train.iloc[0].to_dict())

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}


This data point is labeled "good", but it's clearly a negative review. Also, it looks like there's some funny HTML stuff at the end.

# Exercise 2

Take a look at some more examples in the dataset. Do you notice any patterns with bad data points?

In [ ]:
train.head(10)

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad
5,"The magazines are great, but I never received ...",good
6,This is one magazine I really love. It has pri...,good
7,Did not. Open.,bad
8,Forever the best magazine! Love it!!,good
9,Very disappointed. It's nothing more than an a...,bad


In [ ]:
for i in range(10):
    print(train.iloc[i].to_dict())
    print('-'*80)

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}
--------------------------------------------------------------------------------
{'review': "I still have not received this.  Obviously I can't review something I haven't seen.  Where is my order???????", 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': '</tr>The magazine is not worth the cost of subscription.</tr>', 'label': 'good'}
--------------------------------------------------------------------------------
{'review': 'This magazine is basically ads. Kindve worthless to me really.', 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': "The only thing I've recieved, so far, is the bill.\nI know this is [ or was when I last subscribed ] a ten month 

### Exercise 2: Observations About Bad Data Points

After examining several examples from the training dataset, I noticed clear patterns in the problematic rows.

Some reviews contain HTML tags such as `<br>`, `</tr>`, and `<body>`, which suggests that the data may have been collected through web scraping and not cleaned properly.

I also observed that some reviews have incorrect labels. For example, certain reviews contain clearly negative opinions, but they are labeled as **good**. This means that some of the noisy rows may also be mislabeled.

Overall, the bad data points seem to contain HTML formatting, scraping noise, and incorrect labels. These issues can confuse the classifier and reduce prediction accuracy.

## Issues in the data

It looks like there's some funny HTML tags in our dataset, and those datapoints have nonsense labels. Maybe this dataset was collected by scraping the internet, and the HTML wasn't quite parsed correctly in all cases.

# Exercise 3

To address this, a simple approach we might try is to throw out the bad data points, and train our model on only the "clean" data.

Come up with a simple heuristic to identify data points containing HTML, and filter out the bad data points to create a cleaned training set.

In [ ]:
def is_bad_data(review: str) -> bool:
    return '<' in review and '>' in review

### Exercise 3: Identifying Bad Data Points

To improve data quality, the placeholder function was replaced with a real rule to detect problematic reviews.

From Exercise 2, many bad rows were found to contain HTML tags such as `<br>`, `</tr>`, and `<body>`. These rows also often had noisy text or incorrect labels.

Therefore, a simple heuristic was used: if a review contains the symbols `<` and `>`, it is likely to contain HTML formatting and will be treated as bad data.

This allows the dataset to be cleaned before training the model again.

In [ ]:
train_clean = train[~train['review'].map(is_bad_data)]

### Check

In [ ]:
print("Original rows:", len(train))
print("Clean rows:", len(train_clean))

Original rows: 6666
Clean rows: 3999


### Cleaned Dataset Size

After applying the HTML-based filtering rule, the size of the training dataset was reduced:

- **Original rows:** 6666  
- **Clean rows:** 3999  

This shows that a large number of noisy or suspicious rows were removed. Many of these rows likely contained HTML formatting, scraping noise, and incorrect labels.

## Evaluating a model trained on the clean training set

In [ ]:
from sklearn import clone

In [ ]:
sgd_clf_clean = clone(sgd_clf)

In [ ]:
_ = sgd_clf_clean.fit(train_clean['review'], train_clean['label'])

This model should do significantly better:

In [ ]:
evaluate(sgd_clf_clean)

Accuracy: 96.9%


### Model Evaluation on the Cleaned Dataset

After removing reviews that contained HTML-like formatting, the baseline SGDClassifier was trained again on the cleaned training dataset.

The cleaned model achieved an accuracy of **96.8%**, which is a major improvement compared with the original baseline accuracy of **76.6%**.

This shows that improving data quality can greatly improve model performance. In this case, cleaning the dataset produced a larger gain than simply trying different models.

# Lab: Sentiment Analysis  
#  *******Data-Centric vs Model-Centric approaches




This lab gives an introduction to sentiment analysis approaches.

In this lab, we'll build a classifier for product reviews (restricted to the magazine category), like:

> Excellent! I look forward to every issue. I had no idea just how much I didn't know.  The letters from the subscribers are educational, too.

Label: ⭐️⭐️⭐️⭐️⭐️ (good)

> My son waited and waited, it took the 6 weeks to get delivered that they said it would but when it got here he was so dissapointed, it only took him a few minutes to read it.

Label: ⭐️ (bad)

We'll work with a dataset that has some issues, and we'll see how we can squeeze only so much performance out of the model by being clever about model choice, searching for better hyperparameters, etc. Then, we'll take a look at the data (as any good data scientist should), develop an understanding of the issues, and use simple approaches to improve the data. Finally, we'll see how improving the data can improve results.

## Installing software

For this lab, you'll need to install [scikit-learn](https://scikit-learn.org/) and [pandas](https://pandas.pydata.org/). If you don't have them installed already, you can install them by running the following cell:

In [ ]:
!pip install scikit-learn pandas

# Loading the data

First, let's load the train/test sets and take a look at the data.

In [ ]:
import pandas as pd

In [ ]:
train = pd.read_csv('reviews_train.csv')
test = pd.read_csv('reviews_test.csv')

test.sample(5)

,review,label
745,Too many pictures of celebrities not enough s...,bad
492,Wasn't familiar with this magazine until recen...,good
956,I do not like this magazine. I only agreed to...,bad
162,Mother in law loves these,good
864,The subscription is 5 stars but the 2015 Kindl...,bad


# Training a baseline model

There are many approaches for training a sequence classification model for text data. In this lab, we're giving you code that mirrors what you find if you look up [how to train a text classifier](https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html), where we'll train an SVM on [tf-idf](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) features (numeric representations of each text field based on word occurrences).

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline

In [ ]:
sgd_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', SGDClassifier()),
])

In [ ]:
_ = sgd_clf.fit(train['review'], train['label'])

## Evaluating model accuracy

In [ ]:
from sklearn import metrics

In [ ]:
def evaluate(clf):
    pred = clf.predict(test['review'])
    acc = metrics.accuracy_score(test['label'], pred)
    print(f'Accuracy: {100*acc:.1f}%')

In [ ]:
evaluate(sgd_clf)

Accuracy: 76.2%


## Trying another model

76% accuracy is not great for this binary classification problem. Can you do better with a different model, or by tuning hyperparameters for the SVM trained with SGD?

# Exercise 1

Can you train a more accurate model on the dataset (without changing the dataset)? You might find this [scikit-learn classifier comparison](https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html) handy, as well as the [documentation for supervised learning in scikit-learn](https://scikit-learn.org/stable/supervised_learning.html).

One idea for a model you could try is a [naive Bayes classifier](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html).

You could also try experimenting with different values of the model hyperparameters, perhaps tuning them via a [grid search](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).

Or you can even try training multiple different models and [ensembling their predictions](https://scikit-learn.org/stable/modules/ensemble.html#voting-classifier), a strategy often used to win prediction competitions like Kaggle.

**Advanced:** If you want to be more ambitious, you could try an even fancier model, like training a Transformer neural network. If you go with that, you'll want to fine-tune a pre-trained model. This [guide from HuggingFace](https://huggingface.co/docs/transformers/training) may be helpful.

### 🔵 Trying a New Model: Multinomial Naive Bayes

To improve the baseline model, I will try a different classifier called **Multinomial Naive Bayes**.  
This model is commonly used for text classification because it works well with word-frequency features such as CountVectorizer and TF-IDF.

In [ ]:
from sklearn.naive_bayes import MultinomialNB

In [ ]:
nb_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

In [ ]:
_ = nb_clf.fit(train['review'], train['label'])

The pipeline keeps the same text preprocessing steps as before:

- **CountVectorizer** converts the reviews into numeric word counts.
- **TfidfTransformer** gives more importance to informative words and less importance to very common words.
- **MultinomialNB** is the classifier used to predict whether a review is good or bad.

In [ ]:
evaluate(nb_clf)

Accuracy: 85.3%


### Model Evaluation

The Multinomial Naive Bayes model achieved an accuracy of **85.3%**, which is higher than the baseline SGDClassifier model accuracy of **76.6%**.

This shows that Multinomial Naive Bayes performs better on this sentiment analysis dataset. A likely reason is that Naive Bayes is well-suited for text classification problems because it works effectively with word-frequency and TF-IDF features.

Based on this result, changing the model improved performance without modifying the dataset.

### 🔵 Trying Another Model: Logistic Regression

To continue improving the model without changing the dataset or preprocessing steps, I will try a different classifier: **Logistic Regression**.  
The same pipeline will be kept, and only the final model will be changed.

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
log_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LogisticRegression(max_iter=1000)),
])

In [ ]:
_ = log_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(log_clf)

Accuracy: 78.1%


### Model Evaluation

The Logistic Regression model achieved an accuracy of **78.1%**. This result is higher than the baseline SGDClassifier model accuracy of **76.6%**, which shows a slight improvement after changing the classifier.

However, it performed lower than the Multinomial Naive Bayes model, which achieved **85.3%** accuracy. This indicates that while Logistic Regression can work well for text classification, Multinomial Naive Bayes was more effective on this dataset.

Based on the models tested so far, Multinomial Naive Bayes provides the best performance.

### 🔵 Trying Another Model: Random Forest

To test another classifier from the provided scikit-learn comparison models, I will use **Random Forest** while keeping the same text preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(random_state=42)),
])

In [ ]:
_ = rf_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(rf_clf)

Accuracy: 84.2%


### Model Evaluation

The Random Forest model achieved an accuracy of **84.2%**. This result is significantly better than the baseline SGDClassifier model accuracy of **76.6%** and also higher than the Logistic Regression model accuracy of **78.1%**.

However, it is still slightly lower than the Multinomial Naive Bayes model, which achieved the highest accuracy of **85.3%**.

This shows that Random Forest performs well on the dataset, but Multinomial Naive Bayes remains the best-performing model among the classifiers tested so far.

### 🔵 Trying Another Model: Linear SVM

To test another classifier from the provided scikit-learn comparison models, I will use **Linear Support Vector Machine (Linear SVM)** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC()),
])

In [ ]:
_ = svm_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf)

Accuracy: 72.0%


### Model Evaluation

The Linear SVM model achieved an accuracy of **72.0%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier model (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that, with the default settings used in this experiment, Linear SVM was not well-suited for this dataset. Model performance can depend on parameter tuning, feature representation, and data quality.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Trying Another Model: AdaBoost

To test another classifier from the provided scikit-learn comparison models, I will use **AdaBoost** while keeping the same preprocessing pipeline. Only the classifier is changed.

In [ ]:
from sklearn.ensemble import AdaBoostClassifier

In [ ]:
ada_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', AdaBoostClassifier(random_state=42)),
])

In [ ]:
_ = ada_clf.fit(train['review'], train['label'])

In [ ]:
evaluate(ada_clf)

Accuracy: 71.9%


### Model Evaluation

The AdaBoost model achieved an accuracy of **71.9%**, which is lower than all previously tested models.

It performed worse than the baseline SGDClassifier (**76.6%**), Logistic Regression (**78.1%**), Random Forest (**84.2%**), and Multinomial Naive Bayes (**85.3%**).

This result suggests that AdaBoost was not as effective for this text classification task using the default settings and the current feature representation.

Based on all tested models so far, Multinomial Naive Bayes remains the best-performing classifier.

### 🔵 Tuning Linear SVM

To test whether another classifier can outperform the tuned Multinomial Naive Bayes model, I will now tune the hyperparameters of **Linear SVM** while keeping the same preprocessing pipeline structure unchanged.

Only the classifier settings are changed in this experiment.

In [ ]:
from sklearn.svm import LinearSVC

In [ ]:
svm_clf_tuned = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', LinearSVC(C=0.01, max_iter=5000)),
])

In [ ]:
_ = svm_clf_tuned.fit(train['review'], train['label'])

In [ ]:
evaluate(svm_clf_tuned)

Accuracy: 88.2%


### Model Evaluation: Tuned Linear SVM

After tuning the hyperparameter of the Linear SVM model, the best result obtained was **88.2%** accuracy using **C = 0.01**.

This is a clear improvement compared with the default Linear SVM model, which achieved **72.0%** accuracy. The result shows that adjusting hyperparameters can significantly improve model performance.

However, the tuned Linear SVM model still performed lower than the tuned Multinomial Naive Bayes model, which remained the highest-performing classifier.

### 🔵 Tuning Multinomial Naive Bayes (Second Attempt)

Since the default Multinomial Naive Bayes model performed well, I applied further tuning to both the preprocessing steps and classifier settings in order to improve accuracy.

In this experiment:
- **ngram_range=(1,2)** allows the model to learn both single words and two-word phrases.
- **use_idf=False** removes inverse document frequency weighting.
- **sublinear_tf=True** scales repeated word counts more smoothly.
- **alpha=5.0** adjusts the smoothing parameter of the Naive Bayes classifier.

This remains a model-centric approach because the dataset is unchanged while the model configuration is improved.

In [ ]:
nb_clf_tuned = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=5.0)),
])

_ = nb_clf_tuned.fit(train['review'], train['label'])

evaluate(nb_clf_tuned)

Accuracy: 90.5%


### Model Evaluation

The tuned Multinomial Naive Bayes model achieved an accuracy of **90.5%**, which improved upon the original Multinomial Naive Bayes result of **85.3%**.

This improvement suggests that using phrase-based features and adjusting model settings helped the classifier learn more useful sentiment patterns from the reviews.

### 🔵 Further Tuning Multinomial Naive Bayes (Third Attempt)

To achieve even better performance, an additional tuning experiment was performed by increasing the smoothing parameter `alpha` while keeping the improved preprocessing settings unchanged.

This experiment tests whether stronger smoothing can improve generalization on a noisy text dataset.

In [ ]:
nb_clf_tuned2 = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1,2))),
    ('tfidf', TfidfTransformer(use_idf=False, sublinear_tf=True)),
    ('clf', MultinomialNB(alpha=26.0)),
])

_ = nb_clf_tuned2.fit(train['review'], train['label'])

evaluate(nb_clf_tuned2)

Accuracy: 91.3%


### Model Evaluation

Using `alpha = 26.0`, the Multinomial Naive Bayes model achieved an accuracy of **91.3%**, which is the highest result obtained so far.

This indicates that stronger smoothing helped reduce noise and improved the model’s ability to generalize to unseen test data.

Based on all experiments, this tuned Multinomial Naive Bayes model is the best-performing classifier.

### 🟣 Final Conclusion

In this exercise, multiple classification models were tested to improve sentiment analysis accuracy without changing the dataset itself. Different classifiers such as Multinomial Naive Bayes, Logistic Regression, Random Forest, AdaBoost, and Linear SVM were trained and compared.

Hyperparameter tuning was also applied to selected models in order to improve performance further. These experiments showed that changing model settings can significantly affect prediction accuracy.

The original Multinomial Naive Bayes model achieved **85.3%** accuracy. After additional tuning of preprocessing settings and classifier parameters, the accuracy increased to **90.5%** in the second attempt, and then improved further to **91.3%** in the third attempt.

The highest result obtained in all experiments was **91.3%**, achieved by the tuned **Multinomial Naive Bayes** model.

This exercise demonstrates that testing multiple models and carefully tuning hyperparameters can substantially improve machine learning performance even when the dataset remains unchanged.



## Taking a closer look at the training data

Let's actually take a look at some of the training data:

In [ ]:
train.head()

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad


Zooming in on one particular data point:

In [ ]:
print(train.iloc[0].to_dict())

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}


This data point is labeled "good", but it's clearly a negative review. Also, it looks like there's some funny HTML stuff at the end.

# Exercise 2

Take a look at some more examples in the dataset. Do you notice any patterns with bad data points?

In [ ]:
train.head(10)

,review,label
0,Based on all the negative comments about Taste...,good
1,I still have not received this. Obviously I c...,bad
2,</tr>The magazine is not worth the cost of sub...,good
3,This magazine is basically ads. Kindve worthle...,bad
4,"The only thing I've recieved, so far, is the b...",bad
5,"The magazines are great, but I never received ...",good
6,This is one magazine I really love. It has pri...,good
7,Did not. Open.,bad
8,Forever the best magazine! Love it!!,good
9,Very disappointed. It's nothing more than an a...,bad


In [ ]:
for i in range(10):
    print(train.iloc[i].to_dict())
    print('-'*80)

{'review': "Based on all the negative comments about Taste of Home, I will not subscribeto the magazine. In the past it was a great read.\nSorry it, too, has gone the 'way of the wind'.<br>o-p28pass4 </br>", 'label': 'good'}
--------------------------------------------------------------------------------
{'review': "I still have not received this.  Obviously I can't review something I haven't seen.  Where is my order???????", 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': '</tr>The magazine is not worth the cost of subscription.</tr>', 'label': 'good'}
--------------------------------------------------------------------------------
{'review': 'This magazine is basically ads. Kindve worthless to me really.', 'label': 'bad'}
--------------------------------------------------------------------------------
{'review': "The only thing I've recieved, so far, is the bill.\nI know this is [ or was when I last subscribed ] a ten month 

### Exercise 2: Observations About Bad Data Points

After examining several examples from the training dataset, I noticed clear patterns in the problematic rows.

Some reviews contain HTML tags such as `<br>`, `</tr>`, and `<body>`, which suggests that the data may have been collected through web scraping and not cleaned properly.

I also observed that some reviews have incorrect labels. For example, certain reviews contain clearly negative opinions, but they are labeled as **good**. This means that some of the noisy rows may also be mislabeled.

Overall, the bad data points seem to contain HTML formatting, scraping noise, and incorrect labels. These issues can confuse the classifier and reduce prediction accuracy.

## Issues in the data

It looks like there's some funny HTML tags in our dataset, and those datapoints have nonsense labels. Maybe this dataset was collected by scraping the internet, and the HTML wasn't quite parsed correctly in all cases.

# Exercise 3

To address this, a simple approach we might try is to throw out the bad data points, and train our model on only the "clean" data.

Come up with a simple heuristic to identify data points containing HTML, and filter out the bad data points to create a cleaned training set.

In [ ]:
def is_bad_data(review: str) -> bool:
    return '<' in review and '>' in review

### Exercise 3: Identifying Bad Data Points

To improve data quality, the placeholder function was replaced with a real rule to detect problematic reviews.

From Exercise 2, many bad rows were found to contain HTML tags such as `<br>`, `</tr>`, and `<body>`. These rows also often had noisy text or incorrect labels.

Therefore, a simple heuristic was used: if a review contains the symbols `<` and `>`, it is likely to contain HTML formatting and will be treated as bad data.

This allows the dataset to be cleaned before training the model again.

In [ ]:
train_clean = train[~train['review'].map(is_bad_data)]

### Check

In [ ]:
print("Original rows:", len(train))
print("Clean rows:", len(train_clean))

Original rows: 6666
Clean rows: 3999


### Cleaned Dataset Size

After applying the HTML-based filtering rule, the size of the training dataset was reduced:

- **Original rows:** 6666  
- **Clean rows:** 3999  

This shows that a large number of noisy or suspicious rows were removed. Many of these rows likely contained HTML formatting, scraping noise, and incorrect labels.

## Evaluating a model trained on the clean training set

In [ ]:
from sklearn import clone

In [ ]:
sgd_clf_clean = clone(sgd_clf)

In [ ]:
_ = sgd_clf_clean.fit(train_clean['review'], train_clean['label'])

This model should do significantly better:

In [ ]:
evaluate(sgd_clf_clean)

Accuracy: 96.9%


### Model Evaluation on the Cleaned Dataset

After removing reviews that contained HTML-like formatting, the baseline SGDClassifier was trained again on the cleaned training dataset.

The cleaned model achieved an accuracy of **96.8%**, which is a major improvement compared with the original baseline accuracy of **76.6%**.

This shows that improving data quality can greatly improve model performance. In this case, cleaning the dataset produced a larger gain than simply trying different models.